Added D2 Regression Metrics

anmolsjoshi opened the pull request:
Related to #307 

Description:
* Canberra Metric
* Fractional Absolute Error
* Wave Hedges Distance
* Geometric Mean Absolute Error
* ~~Geometric Mean Relative Absolute Error~~

Check list:
* [x] New tests are added (if a new feature is modified)
* [x] New doc strings: text and/or example code are in RST format
* [x] Documentation is updated (if required)

anmolsjoshi commented:
Should have the tests completed in a few hours!

anmolsjoshi commented:
cc @vfdev-5 @zasdfgbnm

zasdfgbnm commented on ignite/contrib/metrics/regression/geometric_mean_absolute_error.py:
```diff
@@ -0,0 +1,33 @@
+        if self._num_examples == 0:
+            raise NotComputableError('GeometricMeanAbsoluteError must have at'
+                                     'least one example before it can be computed.')
+        return np.power(self._product_of_errors, 1.0 / self._num_examples)
```
I suggest to replace `np.power(a, b)` with `a ** b`, since numpy is not a dependency of ignite.

anmolsjoshi replied:
Agree will incorporate tonight.

zasdfgbnm commented on ignite/contrib/metrics/regression/geometric_mean_relative_absolute_error.py:
```diff
@@ -0,0 +1,39 @@
+        # once num_examples is updated, new average is (prev_sum + y.sum())/num_examples
+        self._mean_of_actuals = (prev_sum_of_gt + y.sum().item()) / self._num_examples
+        relative_error = errors / torch.abs(y - self._mean_of_actuals)
+        self._product_of_errors = self._product_of_errors * torch.prod(relative_error).item()
```
I don't think this metric is possible to implement without saving all the history. Since the A bar is unknown until the last sample come in, it is impossible to compute the denominator with the correct value of A bar. The implementation now looks not correct to me because in `_product_of_errors` is computed using old A bar, rather than the correct value of A bar.

anmolsjoshi replied:
Ok! I’ll remove this from this PR but correct the implementation separately. We can add it once we decided how to handle metics that require the entire history of the training.

anmolsjoshi commented:
@vfdev-5 @zasdfgbnm I have incorporated comments

- removed GMRAE implementation pending further discussion about recording the entire history of training
- changed calculation of power to remove numpy

let me know if there are any issues

anmolsjoshi commented:
cc @vfdev-5

vfdev-5 commented on ignite/contrib/metrics/regression/canberra_metric.py:
```diff
@@ -0,0 +1,26 @@
+
+class CanberraMetric(Metric):
+    r"""
+    Calculates the Canberra Metric.
```
Could you please unify the docstring like for other implementations, e.g
```
    Calculates the Manhattan Distance:
    :math:`\text{MD} = \sum_{j=1}^n (A_j - P_j)`,
    where :math:`A_j` is the ground truth and :math:`P_j` is the predicted value.
    More details can be found in `Botchkarev 2018`__.
    - `update` must receive output of the form `(y_pred, y)`.
    - `y` and `y_pred` must be of same shape.
    __ https://arxiv.org/abs/1809.03006

```

anmolsjoshi replied:
Done!

vfdev-5 commented on ignite/contrib/metrics/regression/wave_hedges_distance.py:
```diff
@@ -0,0 +1,26 @@
+
+    def update(self, output):
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred) / torch.max(y_pred, y.view_as(y_pred))
```
I have a doubt on this one, I'm not sure that $max(A_j, P_j)$ = `torch.max(y_pred, y.view_as(y_pred))`. Need to check this from my side

anmolsjoshi replied:
@vfdev-5 see code below

```python
y_pred = torch.rand(10)
y = torch.rand(10)
maxes = torch.max(y_pred, y)

print (y_pred)
# tensor([0.1823, 0.0923, 0.1578, 0.9375, 0.4769, 0.0928, 0.7895, 0.2848, 0.4338, 0.2930])
print (y)
# tensor([0.3095, 0.8486, 0.2325, 0.3083, 0.6410, 0.8848, 0.5019, 0.0213, 0.1661, 0.2154])
print (maxes)
# tensor([0.3095, 0.8486, 0.2325, 0.9375, 0.6410, 0.8848, 0.7895, 0.2848, 0.4338, 0.2930])
```

vfdev-5 replied:
You are right, `torch.max` is element-wise max if two tensors are compared.

vfdev-5 commented:
@anmolsjoshi thanks for pinging ! I left some comments. Otherwise looks good.

anmolsjoshi commented:
@vfdev-5 thanks for the review. responded to comments.

vfdev-5 commented:
I wonder also whether we should check input data shape ?

anmolsjoshi commented:
yes, I think this is something required in all the regression metrics, currently none of them employ it. It can be as simple as:

```python
if y_pred.shape == y.shape:
    raise ValueError('......')
```

vfdev-5 commented on ignite/contrib/metrics/regression/geometric_mean_absolute_error.py:
```diff
@@ -0,0 +1,37 @@
+    r"""
+    Calculates the Geometric Mean Absolute Error.
+
+    :math:`\text{GMRAE} = \sqrt[n]{\prod_{j=1}^{\n}{|A_j - P_j|}`
```
```suggestion
    :math:`\text{GMRAE} = \sqrt[n]{\prod_{j=1}^{n}{|A_j - P_j|}}`
```

vfdev-5 commented:
That's true that prediction and true values should have the same shape, but shouldn't they be just vectors : `(N, )` ?

vfdev-5 commented on ignite/contrib/metrics/regression/geometric_mean_absolute_error.py:
```diff
@@ -0,0 +1,37 @@
+    def update(self, output):
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred)
+        self._product_of_errors = self._product_of_errors * torch.prod(errors).item()
```
```suggestion
        self._product_of_errors *= torch.prod(errors).item()
```

vfdev-5 commented on ignite/contrib/metrics/regression/geometric_mean_absolute_error.py:
```diff
@@ -0,0 +1,37 @@
+    def update(self, output):
+        y_pred, y = output
+        errors = torch.abs(y.view_as(y_pred) - y_pred)
+        self._product_of_errors = self._product_of_errors * torch.prod(errors).item()
```
I think we should implement this differently (e.g. with log) in order to underflow when multiply error values ~1e-1, 1e-2.

anmolsjoshi replied:
Agree. I have a working local implementation I can push

anmolsjoshi replied:
Added implementation using log.

anmolsjoshi commented:
So would the shape check confirm that y and y_pred are the same shape and (N, ) or (N, 1) ?

anmolsjoshi commented:
@vfdev-5 I updated the implementation of GMAE using log and exp. I left the tests the same to see if I got the same answers and I did. Let me know if you want me to update the test to incorporate the current implementation. 

Also what are your thoughts on creating a `BaseRegressionMetric` class that include a shape check as we discussed above. All regression modules can inherit this. Shape check, checks if y_pred and y are (N) or (N, 1), it squeezes both to (N,).

vfdev-5 commented:
> Also what are your thoughts on creating a BaseRegressionMetric class that include a shape check as we discussed above. All regression modules can inherit this. Shape check, checks if y_pred and y are (N) or (N, 1), it squeezes both to (N,).

@anmolsjoshi seems good.

vfdev-5 commented:
Curiously exp-log implementation of GMAE works without any explosions on zero error:
```python
y_true = torch.Tensor([0.0, 0.0, 0.0, 0.0])
y_pred = torch.Tensor([0.01, 0.0, 0.0001, 0.2])

gmae.update((y_pred, y_true))
gmae._sum_of_errors, gmae.compute()
> tensor(-inf), 0.0
```

vfdev-5 commented:
@anmolsjoshi @jasonkriss please review 

@jasonkriss merge if ok

anmolsjoshi commented:
@vfdev-5 LGTM! Thanks for the last few commits!

jasonkriss commented:
LGTM! Nice work guys!
